File: geeksforgeeks/arrays/contiguous-subarray.go
package main

import (
	"fmt"
)

/*
	Given n (size of array) and s (sum to add up to),
	find a coniguous subarray in which the elements add up to the sum.
	If there is no subarray return -1.
	Ex: n = 5, s = 12
	arr = [1, 2, 3, 7, 5]
	output: [2, 3, 7]
*/

func main() {
	nums := []int{1, 2, 3, 7, 5}
	sum := 12
	arr := contigSubarrayEqToSum(nums, sum)
	fmt.Println(arr)

}

func contigSubarrayEqToSum(nums []int, sum int) []int {
	start := 0
	end := 0

	if len(nums) == 0 {
		return []int{-1}
	}

	contiguousSum := nums[end]
	var contigSubarray []int
	contigSubarray = append(contigSubarray, nums[end])

	for end < len(nums) {
		fmt.Println(contigSubarray)
		if contiguousSum == sum {
			return contigSubarray
		} else if contiguousSum < sum {
			end++
			if end < len(nums) {
				contigSubarray = append(contigSubarray, nums[end])
				contiguousSum += nums[end]
			}
		} else if contiguousSum > sum && start < end {
			start++
			contiguousSum -= contigSubarray[0]
			contigSubarray = contigSubarray[1:]
		}
	}

	return []int{-1}
}

File: geeksforgeeks/arrays/contiguous-subarray.dart
/*
	Given n (size of array) and s (sum to add up to),
	find a coniguous subarray in which the elements add up to the sum.
	If there is no subarray return -1.
	Ex: n = 5, s = 12
	arr = [1, 2, 3, 7, 5]
	output: [2, 3, 7]
*/

main() {
	List nums = [1, 2, 3, 7, 5];
	int sum = 12;
	var arr = contigSubarrayEqToSum(nums, sum);
  print(arr);
}

List contigSubarrayEqToSum(List nums, int sum) {
	int start = 0;
	int end = 0;

	if(nums.length == 0) {
		return [-1];
	}

	int contiguousSum = nums[end];
	List contigSubarray = [nums[end]];

  while(end < nums.length) {
    if(contiguousSum == sum) {
      return contigSubarray;
    } else if(contiguousSum < sum) {
      end++;
      if(end < nums.length) {
        contigSubarray.add(nums[end]);
        contiguousSum += nums[end];
      }
    } else if(contiguousSum > sum && start < end) {
      contiguousSum -= contigSubarray[0];
      contigSubarray.removeAt(0);
    }
  }

	return [-1];
}
